funsor/gaussian.py

```diff
@@ -334,47 +335,56 @@ def eager_subs(self, subs):
         if not subs:
             return self
 
-        # Constants and Variables are eagerly substituted;
+        # Constants and Affine funsors are eagerly substituted;
         # everything else is lazily substituted.
         lazy_subs = tuple((k, v) for k, v in subs
-                          if not isinstance(v, (Number, Tensor, Variable, Slice)))
+                          if not isinstance(v, (Number, Tensor, Variable, Slice))
+                          and not affine_inputs(v))
         var_subs = tuple((k, v) for k, v in subs if isinstance(v, Variable))
         int_subs = tuple((k, v) for k, v in subs if isinstance(v, (Number, Tensor, Slice))
                          if v.dtype != 'real')
         real_subs = tuple((k, v) for k, v in subs if isinstance(v, (Number, Tensor))
                           if v.dtype == 'real')
-        if not (var_subs or int_subs or real_subs):
-            return reflect(Subs, self, lazy_subs)
-
-        # First perform any variable substitutions.
+        affine_subs = tuple((k, v) for k, v in subs
+                            if not isinstance(v, Variable) and affine_inputs(v))
         if var_subs:
-            rename = {k: v.name for k, v in var_subs}
-            inputs = OrderedDict((rename.get(k, k), d) for k, d in self.inputs.items())
-            if len(inputs) != len(self.inputs):
-                raise ValueError("Variable substitution name conflict")
-            var_result = Gaussian(self.info_vec, self.precision, inputs)
-            new_subs = int_subs + real_subs + lazy_subs
-            return Subs(var_result, new_subs) if new_subs else var_result
-
-        # Next perform any integer substitution, i.e. slicing into a batch.
+            return self._eager_subs_var(var_subs, int_subs + real_subs + affine_subs + lazy_subs)
         if int_subs:
-            int_inputs = OrderedDict((k, d) for k, d in self.inputs.items() if d.dtype != 'real')
-            real_inputs = OrderedDict((k, d) for k, d in self.inputs.items() if d.dtype == 'real')
-            tensors = [self.info_vec, self.precision]
-            funsors = [Subs(Tensor(x, int_inputs), int_subs) for x in tensors]
-            inputs = funsors[0].inputs.copy()
-            inputs.update(real_inputs)
-            int_result = Gaussian(funsors[0].data, funsors[1].data, inputs)
-            new_subs = real_subs + lazy_subs
-            return Subs(int_result, new_subs) if new_subs else int_result
-
+            return self._eager_subs_int(int_subs, real_subs + affine_subs + lazy_subs)
+        if real_subs:
+            return self._eager_subs_real(real_subs, affine_subs + lazy_subs)
+        # TODO
```

Comment:
I will implement this in the next PR. I don't think Du Phan would be interested in this PR, so I'll add the interesting stuff in the follow-up PR.